src/frontend/apps/impress/src/features/docs/doc-management/libs/CollaborationProvider.ts

```diff
@@ -0,0 +1,333 @@
+  }
+
+  protected isDuplicateAwareness(message64: string): boolean {
+    if (this.treatedAwarenessMessage === message64) {
+      return true;
+    }
+
+    this.treatedAwarenessMessage = message64;
+
+    return false;
+  }
+
+  /**
+   * Outgoing message event
+   *
+   * Sent to the server the message to
+   * be sent to the other users
+   */
+  public async onPollOutgoingMessage({ message }: onOutgoingMessageParameters) {
+    if (!this.isWebsocketFailed || !this.canEdit) {
+      return;
+    }
+
+    const message64 = Buffer.from(message.toUint8Array()).toString('base64');
+    if (this.isDuplicateAwareness(message64)) {
+      return;
+    }
+
+    try {
+      const { updated } = await pollOutgoingMessageRequest({
+        pollUrl: this.toPollUrl('message'),
+        message64,
+      });
+
+      if (!updated) {
+        await this.pollSync();
+      }
+    } catch (error: unknown) {
+      if (isAPIError(error)) {
+        // The user is not allowed to send messages
+        if (error.status === 403) {
+          this.off('outgoingMessage', this.onPollOutgoingMessage.bind(this));
+          this.canEdit = false;
+        }
+      }
+    }
+  }
+
+  /**
+   * EventSource is a API for opening an HTTP
+   * connection for receiving push notifications
+   * from a server in real-time.
+   * We use it to sync the document with the server
+   */
+  protected initCollaborationSSE() {
+    if (!this.isWebsocketFailed) {
+      return;
+    }
+
+    this.sse = new EventSource(this.toPollUrl('message'), {
+      withCredentials: true,
+    });
+
+    this.sse.onmessage = (event) => {
+      const { updatedDoc64, stateFingerprint, awareness64 } = JSON.parse(
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+        event.data,
+      ) as {
+        updatedDoc64?: string;
+        stateFingerprint?: string;
+        awareness64?: string;
+      };
+
+      if (awareness64) {
+        if (this.isDuplicateAwareness(awareness64)) {
+          return;
+        }
+
+        this.treatedAwarenessMessage = awareness64;
+        const awareness = Buffer.from(awareness64, 'base64');
+
+        this.onMessage({
+          data: awareness,
+        } as MessageEvent);
+      }
+
+      if (updatedDoc64) {
+        this.document.transact(() => {
+          Y.applyUpdate(this.document, Buffer.from(updatedDoc64, 'base64'));
+        }, this);
+      }
+
+      const localStateFingerprint = this.getStateFingerprint(this.document);
+      if (localStateFingerprint !== stateFingerprint) {
+        void this.pollSync();
+      } else {
+        this.seemsUnsyncCount = 0;
+      }
+    };
+
+    this.sse.onopen = () => {};
+
+    this.sse.onerror = (err) => {
+      console.error('SSE error:', err);
+      this.sse?.close();
+
+      setTimeout(() => {
+        this.initCollaborationSSE();
+      }, 5000);
+    };
+  }
+
+  /**
+   * Sync the document with the server.
+   *
+   * In some rare cases, the document may be out of sync.
```

Comment:
Similar to above, could you also call the existing HP `forceSync` and let the Yjs sync protocol handle this? passing Yjs documents and updates around seems a little dangerous to me (at least it's difficult for me to verify if this is correct or not)